### ISSUE TITLE
Z3 QF_S logic

### ISSUE THREAD

**Issue by LeventErkok** [LeventErkok/sbv#204]
Z3 now supports strings/sequences. See if we can take advantage of this. (Mapping Haskell's String type, possibly?)

**Comment by LeventErkok** [comment]
Paper about this work: http://research.microsoft.com/en-us/people/nbjorner/smtlibstr.pdf

**Comment by LeventErkok** [comment]
Comment from Pierre:
 

> I perform a lot of model 2 model transformations, where the final model is represented as a string that should adhere to a grammar or a regular expression. Checking that a program can’t generate anything that violates this grammar or regular expression would be extremely useful. For a regular expression, I am convinced that the current problem solvers can check it. For a grammar, I wonder whether current problem solvers can handle this kind of complexity.

So, any "string" produced will match a given regexp is a good use-case to focus on.

**Comment by LeventErkok** [comment]
Note: It appears Z3 doesn't actually support "symbolic" regular-expressions. So, this could be a limiting factor as we design the API. Something to ponder. See here: https://github.com/Z3Prover/z3/issues/364#issuecomment-185316436

**Comment by LeventErkok** [comment]
@joelburget

I've pulled your string implementation to the `symbolic-strings` branch, and I am currently playing around with it. Once it is in good shape, I'll merge it to master.

Currently, it looks like we need some basic issues:

```
Prelude Data.SBV> sat $ \s -> s .== literal "abc" `strConcat` literal "def"
*** Exception:
*** Data.SBV: Unexpected response from the solver, context: getModel:
***
***    Sent      : (get-value (s0))
***    Expected  : a value binding for kind: SString
***    Received  : ((s0 "abcdef"))
***
***    Executable: z3
***    Options   : -nw -in -smt2
```

I'll look at these, but if you have patches, please directly submit to the `symbolic_strings` branch.

**Comment by LeventErkok** [comment]
Things to do (at least):


* Fix the model extraction (see prev comment)
* Implement concrete optimizations for string ops
* Handle C translation (or at least gracefully fail)
* More tests and examples

**Comment by LeventErkok** [comment]
@joelburget 

This now works:

```
Prelude Data.SBV> sat $ \s -> s .== literal "abc" `strConcat` literal "def"
Satisfiable. Model:
  s0 = "abcdef" :: String
```

Was there a reason why you created a separate `EString` constructor in `SExpr`? I think `ECon` should handle that case just fine, with other functions carefully reconstructing the value based on kind information we have elsewhere. I don't see an explicit reason for `EString`; let me know if I'm missing something here. (Still needs testing of course.)

**Comment by LeventErkok** [comment]
I think I've addressed all the concerns. What remains is more tests and examples before we can merge to master.

Once that is done, we should also consider adding support for regular-expressions before making a general release.

**Comment by joelburget** [comment]
> Was there a reason why you created a separate `EString` constructor in `SExpr`?

No -- as you noticed, I shouldn't have added it.

> I think I've addressed all the concerns.

Great! Thank you for fixing it up :)

> we should also consider adding support for regular-expressions

They would be nice to have, though I don't need them for the work I'm doing. More likely to be useful, for me at least, is sequences (which should be fairly easy since the API looks almost exactly the same as for strings).

**Comment by LeventErkok** [comment]
Actually, I'm about to turn in support for RegExps, which is rather easy.

For sequences: That's a different story since it can be parameterized by an arbitrary symbolic type. I'm not sure how much of a work that would be to support it wholly. Would be good to experiment.

**Comment by LeventErkok** [comment]
This is now all merged to master; and pending a few polishing items is about ready to go.
